warehouse/accounts/utils.py

```diff
@@ -0,0 +1,181 @@
+    # The following string was obtained by:
+    #     base64.b64encode(b"__token__:pypi-").decode("utf-8")
+    # Basic auth is standard base64, so non-alphanumeric chars are + and /
+    pattern = re.compile(r"X190b2tlbl9fOnB5cGkt[A-Za-z0-9+/=]+")
+
+    def extract(self, text):
+        try:
+            _, token = (
+                base64.b64decode(text.encode("utf-8")).decode("utf-8").split(":", 1)
+            )
+            return token
+        except Exception as exc:
+            raise ExtractionFailed from exc
+
+
+TOKEN_LEAK_MATCHERS = {
+    matcher.name: matcher
+    for matcher in [PlainTextTokenLeakMatcher(), Base64BasicAuthTokenLeakMatcher()]
+}
+
+
+class InvalidTokenLeakRequest(Exception):
+    def __init__(self, message, reason):
+        self.reason = reason
+        super().__init__(message)
+
+
+@attr.s(auto_attribs=True, frozen=True, slots=True)
+class TokenLeakDisclosureRequest:
+
+    token: str
+    public_url: str
+
+    @classmethod
+    def from_api_record(cls, record, *, matchers=TOKEN_LEAK_MATCHERS):
+
+        if not isinstance(record, dict):
+            raise InvalidTokenLeakRequest(
+                f"Record is not a dict but: {str(record)[:100]}", reason="format"
+            )
+
+        missing_keys = sorted({"token", "type", "url"} - set(record))
+        if missing_keys:
+            raise InvalidTokenLeakRequest(
+                f"Record is missing attribute(s): {', '.join(missing_keys)}",
+                reason="format",
+            )
+
+        matcher_code = record["type"]
+
+        matcher = matchers.get(matcher_code)
+        if not matcher:
+            raise InvalidTokenLeakRequest(
+                f"Matcher with code {matcher_code} not found. "
+                f"Available codes are: {', '.join(matchers)}",
+                reason="invalid_matcher",
+            )
+
+        try:
+            extracted_token = matcher.extract(record["token"])
+        except ExtractionFailed:
+            raise InvalidTokenLeakRequest(
+                f"Cannot extract token from recieved match", reason="extraction"
+            )
+
+        return cls(token=extracted_token, public_url=record["url"])
+
+
+class TokenLeakAnalyzer:
+    def __init__(self, request):
+        self._request = request
+        self._metrics = self._request.find_service(IMetricsService, context=None)
+        self._macaroon_service = self._request.find_service(
+            IMacaroonService, context=None
+        )
+
+    def analyze_disclosure(self, disclosure_record, origin):
+
+        self._metrics.increment(f"warehouse.token_leak.{origin}.recieved")
+
+        try:
+            disclosure = TokenLeakDisclosureRequest.from_api_record(
+                record=disclosure_record
+            )
+        except InvalidTokenLeakRequest as exc:
+            # TODO Logging something here would be useful in case we recieve
+            # unexpected pattern type.
```

Comment:
We don't really rely on logs very much. Ideally we would be watching the metrics for the types of exceptions occurring here instead.

If this is truly unexpected and "shouldn't happen", we might want to let this exception bubble up into our exception handling service instead, where we'd be immediately notified, and have a lot more context than the metrics service or a log message would be able to provide.